// mdx-m3-viewer/dist/cjs/parsers/mdlx/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const model_1 = require("./model");
const sequence_1 = require("./sequence");
const material_1 = require("./material");
const layer_1 = require("./layer");
const texture_1 = require("./texture");
const textureanimation_1 = require("./textureanimation");
const geoset_1 = require("./geoset");
const geosetanimation_1 = require("./geosetanimation");
const bone_1 = require("./bone");
const light_1 = require("./light");
const helper_1 = require("./helper");
const attachment_1 = require("./attachment");
const particleemitter_1 = require("./particleemitter");
const particleemitter2_1 = require("./particleemitter2");
const particleemitterpopcorn_1 = require("./particleemitterpopcorn");
const ribbonemitter_1 = require("./ribbonemitter");
const camera_1 = require("./camera");
const eventobject_1 = require("./eventobject");
const collisionshape_1 = require("./collisionshape");
const unknownchunk_1 = require("./unknownchunk");
const tokenstream_1 = require("./tokenstream");
const isformat_1 = require("./isformat");
exports.default = {
    Model: model_1.default,
    Sequence: sequence_1.default,
    Material: material_1.default,
    Layer: layer_1.default,
    Texture: texture_1.default,
    TextureAnimation: textureanimation_1.default,
    Geoset: geoset_1.default,
    GeosetAnimation: geosetanimation_1.default,
    Bone: bone_1.default,
    Light: light_1.default,
    Helper: helper_1.default,
    Attachment: attachment_1.default,
    ParticleEmitter: particleemitter_1.default,
    ParticleEmitter2: particleemitter2_1.default,
    ParticleEmitterPopcorn: particleemitterpopcorn_1.default,
    RibbonEmitter: ribbonemitter_1.default,
    Camera: camera_1.default,
    EventObject: eventobject_1.default,
    CollisionShape: collisionshape_1.default,
    UnknownChunk: unknownchunk_1.default,
    TokenStream: tokenstream_1.default,
    isMdx: isformat_1.isMdx,
    isMdl: isformat_1.isMdl,
};
