// mdx-m3-viewer/dist/cjs/parsers/w3x/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const map_1 = require("./map");
const doo_1 = require("./doo");
const imp_1 = require("./imp");
const mmp_1 = require("./mmp");
const shd_1 = require("./shd");
const w3c_1 = require("./w3c");
const w3d_1 = require("./w3d");
const w3e_1 = require("./w3e");
const w3i_1 = require("./w3i");
const w3o_1 = require("./w3o");
const w3r_1 = require("./w3r");
const w3s_1 = require("./w3s");
const w3u_1 = require("./w3u");
const wct_1 = require("./wct");
const wpm_1 = require("./wpm");
const wtg_1 = require("./wtg");
const wts_1 = require("./wts");
const unitsdoo_1 = require("./unitsdoo");
const w3f_1 = require("./w3f");
exports.default = {
    Map: map_1.default,
    doo: doo_1.default,
    imp: imp_1.default,
    mmp: mmp_1.default,
    shd: shd_1.default,
    w3c: w3c_1.default,
    w3d: w3d_1.default,
    w3e: w3e_1.default,
    w3i: w3i_1.default,
    w3o: w3o_1.default,
    w3r: w3r_1.default,
    w3s: w3s_1.default,
    w3u: w3u_1.default,
    wct: wct_1.default,
    wpm: wpm_1.default,
    wtg: wtg_1.default,
    wts: wts_1.default,
    unitsdoo: unitsdoo_1.default,
    w3f: w3f_1.default,
};
