// mdx-m3-viewer/dist/cjs/utils/jass2/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JassTexMapFlags = exports.JassBlendMode = exports.JassRarityControl = exports.JassStartLocPrio = exports.JassPlacement = exports.JassPlayerColor = exports.JassCameraSetup = exports.JassCameraField = exports.JassVolumeGroup = exports.JassPlayerSlotState = exports.JassMapControl = exports.JassMapDensity = exports.JassMapSetting = exports.JassMapVisibility = exports.JassMapFlag = exports.JassGameType = exports.JassGameDifficulty = exports.JassGameSpeed = exports.JassUnitType = exports.JassDialogEvent = exports.JassWidgetEvent = exports.JassLimitOp = exports.JassUnitEvent = exports.JassPlayerUnitEvent = exports.JassPlayerEvent = exports.JassGameEvent = exports.JassEventId = exports.JassAiDifficulty = exports.JassUnitState = exports.JassPlayerGameResult = exports.JassPlayerScore = exports.JassPlayerState = exports.JassFGameState = exports.JassIGameState = exports.JassGameState = exports.JassRacePreference = exports.JassAllianceType = exports.JassRace = exports.JassRect = exports.JassRegion = exports.JassLocation = exports.JassTimer = exports.JassTrigger = exports.JassGroup = exports.JassForce = exports.JassUnit = exports.JassWidget = exports.JassPlayer = exports.JassAgent = exports.JassHandle = void 0;
exports.JassHashtable = exports.JassPathingType = exports.JassSoundType = exports.JassWeaponType = exports.JassDamageType = exports.JassAttackType = exports.JassItemType = exports.JassVersion = exports.JassFogState = exports.JassWeatherEffect = exports.JassEffectType = void 0;
const handle_1 = require("./handle");
exports.JassHandle = handle_1.default;
const agent_1 = require("./agent");
exports.JassAgent = agent_1.default;
// import JassEvent from './event';
const player_1 = require("./player");
exports.JassPlayer = player_1.default;
const widget_1 = require("./widget");
exports.JassWidget = widget_1.default;
const unit_1 = require("./unit");
exports.JassUnit = unit_1.default;
// import JassDestructable from './destructable';
// import JassItem from './item';
// import JassAbility from './ability';
// import JassBuff from './buff';
const force_1 = require("./force");
exports.JassForce = force_1.default;
const group_1 = require("./group");
exports.JassGroup = group_1.default;
const trigger_1 = require("./trigger");
exports.JassTrigger = trigger_1.default;
// import JassTriggerCondition from './triggercondition';
// import JassTriggerAction from './triggeraction';
const timer_1 = require("./timer");
exports.JassTimer = timer_1.default;
const location_1 = require("./location");
exports.JassLocation = location_1.default;
const region_1 = require("./region");
exports.JassRegion = region_1.default;
const rect_1 = require("./rect");
exports.JassRect = rect_1.default;
// import JassBoolexpr from './boolexpr';
// import JassSound from './sound';
// import JassConditionFunc from './conditionfunc';
// import JassFilterFunc from './filterfunc';
// import JassUnitPool from './unitpool';
// import JassItemPool from './itempool';
const race_1 = require("./race");
exports.JassRace = race_1.default;
const alliancetype_1 = require("./alliancetype");
exports.JassAllianceType = alliancetype_1.default;
const racepreference_1 = require("./racepreference");
exports.JassRacePreference = racepreference_1.default;
const gamestate_1 = require("./gamestate");
exports.JassGameState = gamestate_1.default;
const igamestate_1 = require("./igamestate");
exports.JassIGameState = igamestate_1.default;
const fgamestate_1 = require("./fgamestate");
exports.JassFGameState = fgamestate_1.default;
const playerstate_1 = require("./playerstate");
exports.JassPlayerState = playerstate_1.default;
const playerscore_1 = require("./playerscore");
exports.JassPlayerScore = playerscore_1.default;
const playergameresult_1 = require("./playergameresult");
exports.JassPlayerGameResult = playergameresult_1.default;
const unitstate_1 = require("./unitstate");
exports.JassUnitState = unitstate_1.default;
const aidifficulty_1 = require("./aidifficulty");
exports.JassAiDifficulty = aidifficulty_1.default;
const eventid_1 = require("./eventid");
exports.JassEventId = eventid_1.default;
const gameevent_1 = require("./gameevent");
exports.JassGameEvent = gameevent_1.default;
const playerevent_1 = require("./playerevent");
exports.JassPlayerEvent = playerevent_1.default;
const playerunitevent_1 = require("./playerunitevent");
exports.JassPlayerUnitEvent = playerunitevent_1.default;
const unitevent_1 = require("./unitevent");
exports.JassUnitEvent = unitevent_1.default;
const limitop_1 = require("./limitop");
exports.JassLimitOp = limitop_1.default;
const widgetevent_1 = require("./widgetevent");
exports.JassWidgetEvent = widgetevent_1.default;
const dialogevent_1 = require("./dialogevent");
exports.JassDialogEvent = dialogevent_1.default;
const unittype_1 = require("./unittype");
exports.JassUnitType = unittype_1.default;
const gamespeed_1 = require("./gamespeed");
exports.JassGameSpeed = gamespeed_1.default;
const gamedifficulty_1 = require("./gamedifficulty");
exports.JassGameDifficulty = gamedifficulty_1.default;
const gametype_1 = require("./gametype");
exports.JassGameType = gametype_1.default;
const mapflag_1 = require("./mapflag");
exports.JassMapFlag = mapflag_1.default;
const mapvisibility_1 = require("./mapvisibility");
exports.JassMapVisibility = mapvisibility_1.default;
const mapsetting_1 = require("./mapsetting");
exports.JassMapSetting = mapsetting_1.default;
const mapdensity_1 = require("./mapdensity");
exports.JassMapDensity = mapdensity_1.default;
const mapcontrol_1 = require("./mapcontrol");
exports.JassMapControl = mapcontrol_1.default;
const playerslotstate_1 = require("./playerslotstate");
exports.JassPlayerSlotState = playerslotstate_1.default;
const volumegroup_1 = require("./volumegroup");
exports.JassVolumeGroup = volumegroup_1.default;
const camerafield_1 = require("./camerafield");
exports.JassCameraField = camerafield_1.default;
const camerasetup_1 = require("./camerasetup");
exports.JassCameraSetup = camerasetup_1.default;
const playercolor_1 = require("./playercolor");
exports.JassPlayerColor = playercolor_1.default;
const placement_1 = require("./placement");
exports.JassPlacement = placement_1.default;
const startlocprio_1 = require("./startlocprio");
exports.JassStartLocPrio = startlocprio_1.default;
const raritycontrol_1 = require("./raritycontrol");
exports.JassRarityControl = raritycontrol_1.default;
const blendmode_1 = require("./blendmode");
exports.JassBlendMode = blendmode_1.default;
const texmapflags_1 = require("./texmapflags");
exports.JassTexMapFlags = texmapflags_1.default;
// import JassEffect from './effect';
const effecttype_1 = require("./effecttype");
exports.JassEffectType = effecttype_1.default;
const weathereffect_1 = require("./weathereffect");
exports.JassWeatherEffect = weathereffect_1.default;
// import JassTerrainDeformation from './terraindeformation';
const fogstate_1 = require("./fogstate");
exports.JassFogState = fogstate_1.default;
// import JassFogModifier from './fogmodifier';
// import JassDialog from './dialog';
// import JassButton from './button';
// import JassQuest from './quest';
// import JassQuestItem from './questitem';
// import JassDefeatCondition from './defeatcondition';
// import JassTimerDialog from './timerdialog';
// import JassLeaderboard from './leaderboard';
// import JassMultiboard from './multiboard';
// import JassMultiboardItem from './multiboarditem';
// import JassTrackable from './trackable';
// import JassGameCache from './gamecache';
const version_1 = require("./version");
exports.JassVersion = version_1.default;
const itemtype_1 = require("./itemtype");
exports.JassItemType = itemtype_1.default;
// import JassTextTag from './texttag';
const attacktype_1 = require("./attacktype");
exports.JassAttackType = attacktype_1.default;
const damagetype_1 = require("./damagetype");
exports.JassDamageType = damagetype_1.default;
const weapontype_1 = require("./weapontype");
exports.JassWeaponType = weapontype_1.default;
const soundtype_1 = require("./soundtype");
exports.JassSoundType = soundtype_1.default;
// import JassLightning from './lightning';
const pathingtype_1 = require("./pathingtype");
exports.JassPathingType = pathingtype_1.default;
// import JassImage from './image';
// import JassUbersplat from './ubersplat';
const hashtable_1 = require("./hashtable");
exports.JassHashtable = hashtable_1.default;
