// mdx-m3-viewer/dist/cjs/parsers/w3x/w3i/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const file_1 = require("./file");
const player_1 = require("./player");
const force_1 = require("./force");
const upgradeavailabilitychange_1 = require("./upgradeavailabilitychange");
const techavailabilitychange_1 = require("./techavailabilitychange");
const randomunittable_1 = require("./randomunittable");
const randomunit_1 = require("./randomunit");
const randomitemtable_1 = require("./randomitemtable");
const randomitemset_1 = require("./randomitemset");
const randomitem_1 = require("./randomitem");
exports.default = {
    File: file_1.default,
    Player: player_1.default,
    Force: force_1.default,
    UpgradeAvailabilityChange: upgradeavailabilitychange_1.default,
    TechAvailabilityChange: techavailabilitychange_1.default,
    RandomUnitTable: randomunittable_1.default,
    RandomUnit: randomunit_1.default,
    RandomItemTable: randomitemtable_1.default,
    RandomItemSet: randomitemset_1.default,
    RandomItem: randomitem_1.default,
};
